### DISCUSSION TITLE
Most efficient way of using json (via class or directly?)

### DISCUSSION THREAD

**Discussion by mashedPotatoes12334** [simdjson/simdjson#1557] [Q&A] [answered]
Hi! Below is a scenario I'm a bit curious about that was not listed in the docs.

Suppose I want to serialize some data to `test.json`. The data could be formatted like this:
```
struct Data {
    int val1;
    bool flag;
    int val2;
};
```

Then, suppose I wanted a way to access that data from `test.json`, and store it for application-wide use.

There are a few ways I can think of:

1. Read the file into a `char*`, and then use a `ondemand::parser` to parse the file. Then, get the parser's iterator. Then, use the `[]` operator to access each individual field of `test.json`, and assign it to each field of `Data`. Then, you just access `Data` if you want to get those values.
2. Do the same thing as step 1, but instead of a `ondemand::parser`, use a `dom::parser`.
3. Use a `dom::parser` or `ondemand::parser` directly. So instead of having to read each field of the json into a `Data` class, just use the parser directly. For example, instead of `data.val1 = parser["val1"];`, just use `parser["val1"]` every time you want to get a field. (Storing the parser instance somewhere)

What is the most performant method?

**Comment by lemire** [comment, marked as answer]
It is going to be very hard to beat the On Demand -> Your Own Struct approach and that is what we recommend.

No generic DOM data structure can beat `Data.val1`. It is a direct access. The DOM data structure will almost always lead to some overhead. No DOM data structure can use less memory.

**Reply by mashedPotatoes12334** [reply]
Ah ok. That makes sense.

Would it be more performant to use `ondemand::parser` or `dom::parser` if you were reading it into a struct? For example, if there were 3 fields, and you did 
```
data.val1 = ondemandParser["val1"];
data.flag = ondemandParser["flag"];
data.val2 = ondemandParser["val2"];
```
Would it be better to use `dom` or `ondemand`, just to clarify?

**Reply by lemire** [reply]
`ondemand` would be faster. That's why we are promoting it as our new default.

Even if you do not care about speed per se, it will be generally more efficient (less memory, less energy use).

**Reply by mashedPotatoes12334** [reply]
Ah ok. Thank you!
